batch-processing/pipeline.go

```diff
@@ -0,0 +1,353 @@
+package main
+
+import (
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const (
+	READ     = "read"
+	DOWNLOAD = "download"
+	CONVERT  = "convert"
+	UPLOAD   = "upload"
+	WRITE    = "write"
+)
+
+var TASKS = []string{DOWNLOAD, CONVERT, UPLOAD}
+
+var (
+	OutputHeader       = []string{"url", "input", "output", "s3url", "error"}
+	FailedOutputHeader = []string{"url"}
+)
+
+var (
+	ErrDuplicateURL    = errors.New("duplicate URL, skipping")
+	ErrImageExistsInS3 = errors.New("image already exists in S3, skipping")
+)
+
+// Pipeline struct
+type Pipeline struct {
+	config     *Config
+	workers    map[string]int
+	channels   map[string]chan *Out
+	maxRetries uint64    // Max number of retries for downloading and uploading an image
+	urlMap     *sync.Map // this will record urls that are already processed so we don't download again.
+}
+
+type Task func(wg *sync.WaitGroup)
+
+func (p *Pipeline) Read(csvReader *csv.Reader) {
+	// Read the input CSV file
+	// For each line, send the URL to the download channel
+	defer close(p.channels[READ])
+
+	csvReader.FieldsPerRecord = 1
+
+	for {
+		row, err := csvReader.Read()
+
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			p.channels[READ] <- &Out{Err: err}
+			continue
+		}
+		p.channels[READ] <- &Out{Url: row[0]}
+	}
+	log.Println("Read done")
+}
+
+func (p *Pipeline) Download(wg *sync.WaitGroup) {
+	// Download the image from the URL
+	// Send the image to the convert channel
+	defer wg.Done()
+	var (
+		row  *Out
+		ok   bool
+		body io.Reader
+		hash string
+		ext  string
+		err  error
+	)
+	for {
+
+		row, ok = <-p.channels[READ]
+		if !ok {
+			break
+		}
+		if row.Err != nil {
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		// Check if the URL has already been processed
+		// If it has, skip it
+		_, ok := p.urlMap.Load(row.Url)
+		if ok {
+			p.channels[DOWNLOAD] <- &Out{
+				Url: row.Url,
+				Err: ErrDuplicateURL,
+			}
+			continue
+		} else {
+			p.urlMap.Store(row.Url, true)
+		}
+
+		body, ext, hash, err = DownloadWithBackoff(row.Url, p.maxRetries)
+
+		if err != nil {
+			row.Err = err
+			p.channels[DOWNLOAD] <- row
+			continue
+		}
+
+		key := fmt.Sprintf("%s-converted.%s", hash, ext)
+
+		// Check if the image hash is already in the S3 bucket
```

Comment:
I think the hash you're using here is of the _input_ file, right?

Can you compare/contrast the benefits of using the input file or the output file's hash as your uniqueness identifier? (Both make sense, but with different trade-offs :))

Reply:
So the reason I went with input hash is that I know it sooner, I don't need to download, convert and then realise I don't need to process the file.

The downside is that it could be a bit misleading since S3 key doesn't match with the checksum of the file. I would say relying on checksum in the filename shouldn't be used in practice anyways, as s3 offers storing checksums in metadata which will be more reliable.

We can't do searches by metadata, which is needed for duplicate checking.

Reply:
> So the reason I went with input hash is that I know it sooner, I don't need to download, convert and then realise I don't need to process the file.

Yeah, it's around what work we can avoid, and what we trade-off.

If you hash the URL, you can avoid even the download, but your results may be incorrect if the URL changes its content.

If you hash the input file, you can avoid wasting CPU on the conversion, but if people make requests from different inputs which lead to the same output (e.g. someone converts a colour photo, and someone else converts an already-greyscale of the same photo), you may end up storing duplicates in S3, wasting storage space.

If you hash the output file, you can avoid wasting storage space in S3, at the expense of possibly wasting some time doing the conversion before finding out the input images were the same.

You can also _combine_ these to try to get the best of both words (e.g. encode both the input and output hash as metadata), depending on how important those savings are.

> The downside is that it could be a bit misleading since S3 key doesn't match with the checksum of the file. I would say relying on checksum in the filename shouldn't be used in practice anyways, as s3 offers storing checksums in metadata which will be more reliable.
>
> We can't do searches by metadata, which is needed for duplicate checking.

Yeah, in real life if this _actually_ mattered to you you'd almost certainly use some kind of database, but the "what can I encode in a filename" can be a cute way to encode _some_ information without needing to stand up separate infrastructure.